test/mechanics/cell/SimpleAssembler.cpp

```diff
@@ -0,0 +1,96 @@
+#include "BoostUnitTest.h"
+#include <fakeit.hpp>
+#include "mechanics/cell/SimpleAssember.h"
+#include "mechanics/cell/CellInterface.h"
+
+fakeit::Mock<NuTo::CellInterface> MockCell(const NuTo::DofType& dof, Eigen::Vector3i numbering)
+{
+    fakeit::Mock<NuTo::CellInterface> cell;
+
+    NuTo::DofVector<int> mockNumbering;
+    NuTo::DofVector<double> mockGradient;
+    NuTo::DofMatrix<double> mockHessian;
+
+    mockNumbering[dof] = numbering;
+    mockGradient[dof] = Eigen::Vector3d(11, 22, 33);
+    mockHessian(dof, dof).resize(3, 3);
+    mockHessian(dof, dof) << 11, 12, 13, 21, 22, 23, 31, 32, 33;
+
+    Method(cell, DofNumbering) = mockNumbering;
+    Method(cell, Gradient) = mockGradient;
+    Method(cell, Hessian0) = mockHessian;
+
+    return cell;
+}
+
+NuTo::SimpleAssembler SetupAssembler(const NuTo::DofType& d)
+{
+    NuTo::DofContainer<int> numActive;
+    NuTo::DofContainer<int> numDependent;
+
+    numActive[d] = 3;
+    numDependent[d] = 2;
+
+    return NuTo::SimpleAssembler(numActive, numDependent);
+}
+
+BOOST_AUTO_TEST_CASE(AssemberGradient)
+{
+    NuTo::DofType d("0", 1);
+    NuTo::SimpleAssembler assembler = SetupAssembler(d);
+
+    auto mockCell0 = MockCell(d, Eigen::Vector3i(0, 1, 2));
+    auto mockCell1 = MockCell(d, Eigen::Vector3i(2, 3, 4));
+
+    /*
+     *    |  11  22  33  ||
+     *  +          | 11  ||  22  33 |  
+     *                   ||
+     *                   ||
+     *       11  22  44  ||  22  33                
+     *      active       ||  dependent
+     */                  
+    
+
+    NuTo::GlobalDofVector gradient = assembler.BuildGradient({&mockCell0.get(), &mockCell1.get()}, {&d});
+
+    BoostUnitTest::CheckEigenMatrix(gradient.J[d], Eigen::Vector3d(11, 22, 44));
+    BoostUnitTest::CheckEigenMatrix(gradient.K[d], Eigen::Vector2d(22, 33));
+}
+
+BOOST_AUTO_TEST_CASE(AssemberHessian)
+{
+    NuTo::DofType d("0", 1);
+    NuTo::SimpleAssembler assembler = SetupAssembler(d);
+
+    auto mockCell0 = MockCell(d, Eigen::Vector3i(0, 1, 2));
+    auto mockCell1 = MockCell(d, Eigen::Vector3i(2, 3, 4));
+
+
+    /*        
+     *   11    12    13 ||
+     *                  ||
+     *   21    22    23 ||             active J
+     *                  ||
+     *   31    32    44 || 12    13
+     *   ===========================
+     *               21 || 22    23
+     *                  ||             dependent K
+     *               31 || 32    33
+     *
+     *     active J         dependent K
+     */
```

Comment:
Nice diagrams!

Reply:
How do we want to use the GlobalDofMatrix in the time integration scheme? If we want to avoid copying/moving the ii part of the matrix, we will have to expose this matrix somehow. This would then invalidate this GlobalDofMatrix (since we only change the ii component and use this afterwards.  Wouldn't it be possible, not to return a GlobalDofMatrix that itself has the 4 submatrices as members, but directly a tuple of these 4 matrices, thus avoiding a GlobalDofMatrix type at all. In addition, what would be the disadvantage if assembling everything into the complete matrix and then use the block operations of eigen to access the submatrices?

Reply:
This `GlobalDofMatrixSparse` is quite equivalent to the currently used `StructureOutputBlockMatrix`, except that it uses the `DofType` instead of the `enum class eDof`. 
And its usage in the time integration scheme should also be equivalent. 

I introduced a type (instead tuple/vector/array) mainly because the access via JJ, JK, KJ, KK is better than via index. Is [2] == JK or KJ? On top, you can later define handy methods like a custom outstream or arithmetic operations.